packages/cli/src/controllers/oauth/abstract-oauth.controller.ts

```diff
@@ -118,33 +125,67 @@ export abstract class AbstractOAuthController {
 		return await this.credentialsRepository.findOneBy({ id: credentialId });
 	}
 
-	createCsrfState(credentialsId: string): [string, string] {
+	createCsrfState(credentialsId: string, userId?: string): [string, string] {
 		const token = new Csrf();
 		const csrfSecret = token.secretSync();
 		const state: CsrfStateParam = {
 			token: token.create(csrfSecret),
 			cid: credentialsId,
+			createdAt: Date.now(),
+			userId,
 		};
 		return [csrfSecret, Buffer.from(JSON.stringify(state)).toString('base64')];
 	}
 
-	protected decodeCsrfState(encodedState: string): CsrfStateParam {
+	protected decodeCsrfState(encodedState: string, req: AuthenticatedRequest): CsrfStateParam {
 		const errorMessage = 'Invalid state format';
 		const decoded = jsonParse<CsrfStateParam>(Buffer.from(encodedState, 'base64').toString(), {
 			errorMessage,
 		});
+
 		if (typeof decoded.cid !== 'string' || typeof decoded.token !== 'string') {
 			throw new ApplicationError(errorMessage);
 		}
+
+		if (decoded.userId !== req.user?.id) {
+			throw new AuthError('Unauthorized');
+		}
+
 		return decoded;
 	}
 
 	protected verifyCsrfState(decrypted: ICredentialDataDecryptedObject, state: CsrfStateParam) {
 		const token = new Csrf();
-		return (
-			decrypted.csrfSecret === undefined ||
-			!token.verify(decrypted.csrfSecret as string, state.token)
+		if (decrypted.csrfSecret === undefined) return false;
+		if (!token.verify(decrypted.csrfSecret as string, state.token)) return false;
```

Comment:
The type for `ICredentialDataDecrypedObject` needs some tightening. Maybe until then we can type this more narrowly here?

```ts
protected verifyCsrfState(
		decrypted: ICredentialDataDecryptedObject & { csrfSecret?: string },
		state: CsrfStateParam,
	) {
```

If so we can remove `as string` from L160

Reply:
`ICredentialDataDecrypedObject` itself needs to go away. It's not a very useful type, and creates more issue for strong typing than it solves.
I'll try to remove it from all oauth code.